src/protocolsupportpocketstuff/hacks/playerheads/PlayerHeadsPacketListener.java

```diff
@@ -0,0 +1,617 @@
+
+		if (tag != null) {
+			String url = getUrlFromSkull(tag, false);
+
+			if (url != null) {
+				cachedSkullBlock.url = url;
+				cachedSkullBlock.tag = tag;
+			}
+		}
+
+		cachedSkullBlocks.put(position.asLong(), cachedSkullBlock);
+
+		if (!isSpawned)
+			return;
+
+		if (!cachedSkullBlock.isCustomSkull())
+			return;
+
+		if (cachedSkullBlock.isSpawned) {
+			cachedSkullBlock.destroy(this);
+		}
+
+		cachedSkullBlock.spawn(this);
+	}
+
+	protected static byte[] toData(BufferedImage skin) {
+		Validate.isTrue(skin.getWidth() == 64, "Must be 64 pixels wide");
+		Validate.isTrue((skin.getHeight() == 64) || (skin.getHeight() == 32), "Must be 64 or 32 pixels high");
+		ByteArrayOutputStream stream = new ByteArrayOutputStream();
+		for (int y = 0; y < skin.getHeight(); y++) {
+			for (int x = 0; x < skin.getWidth(); x++) {
+				Color color = new Color(skin.getRGB(x, y), true);
+				stream.write(color.getRed());
+				stream.write(color.getGreen());
+				stream.write(color.getBlue());
+				stream.write(color.getAlpha());
+			}
+		}
+		return stream.toByteArray();
+	}
+
+	static class CachedSkullBlock {
+		private long entityId = new SplittableRandom().nextLong(Integer.MAX_VALUE, Long.MAX_VALUE);
+		private Position position;
+		private NBTTagCompoundWrapper tag;
+		private String url;
+		private int dataValue = 1;
+		private boolean isSpawned = false;
+
+		public CachedSkullBlock(Position position) {
+			this.position = position;
+		}
+
+		public boolean isCustomSkull() {
+			return tag != null && url != null;
+		}
+
+		public long getEntityId() {
+			return entityId;
+		}
+
+		public void spawn(PlayerHeadsPacketListener listener) {
+			isSpawned = true;
+
+			if (Skins.INSTANCE.hasPeSkin(url)) {
+				sendFakePlayer(listener, url, Skins.INSTANCE.getPeSkin(url));
+			} else {
+				new Thread() {
+					public void run() {
+						try {
+							BufferedImage image = ImageIO.read(new URL(url));
+							byte[] data = toData(image);
+							sendFakePlayer(listener, url, data);
+							Skins.INSTANCE.cachePeSkin(url, data);
+						} catch (IOException e) {
+							e.printStackTrace();
+						}
+					}
+				}.start();
+			}
+		}
+
+		private void sendFakePlayer(PlayerHeadsPacketListener listener, String url, byte[] data) {
+			int x = tag.getIntNumber("x");
+			int y = tag.getIntNumber("y");
+			int z = tag.getIntNumber("z");
+
+			CollectionsUtils.ArrayMap<DataWatcherObject<?>> metadata = new CollectionsUtils.ArrayMap<>(76);
+
+			metadata.put(39, new DataWatcherObjectFloatLe(1.05f)); // scale, needs to be a *bit* bigger than the original skull
+			metadata.put(54, new DataWatcherObjectFloatLe(0.001f)); // bb width
+			metadata.put(55, new DataWatcherObjectFloatLe(0.001f)); // bb height
+
+			UUID uuid = UUID.randomUUID();
+
+			float yaw = 0;
+			float xOffset = 0.5F;
+			float yOffset = 0.0F;
+			float zOffset = 0.5F;
+
+			if (dataValue == 1) { // on ground
+				int rot = tag.getByteNumber("Rot");
+				switch (rot) {
```

Comment:
Seperate method that retirns rotation based on thingy. Might be helpful in the future